File: apps/desktop/src/api/googleBatch.ts
/**
 * Google Batch API Client
 *
 * Provides TypeScript interface for asynchronous large-volume LLM processing
 * at 50% cost savings with 24-hour SLO.
 */

import { invoke } from '@tauri-apps/api/core';

// ========================================
// Types
// ========================================

export enum BatchJobState {
  PENDING = 'PENDING',
  RUNNING = 'RUNNING',
  SUCCEEDED = 'SUCCEEDED',
  FAILED = 'FAILED',
  CANCELLED = 'CANCELLED',
  EXPIRED = 'EXPIRED',
}

export interface BatchJobStats {
  totalRequests: number;
  completedRequests: number;
  failedRequests: number;
  pendingRequests: number;
  totalTokens?: number;
  totalCost?: number;
}

export interface BatchJobError {
  code: number;
  message: string;
  details?: any[];
}

export interface BatchResult {
  customId?: string;
  index: number;
  response?: any;
  error?: BatchJobError;
}

export interface BatchJob {
  name: string;
  displayName?: string;
  state: BatchJobState;
  model: string;
  createTime: string;
  updateTime?: string;
  endTime?: string;
  stats?: BatchJobStats;
  error?: BatchJobError;
  results?: BatchResult[];
  outputFileUri?: string;
  metadata?: any;
}

export interface ListBatchJobsResponse {
  batchJobs: BatchJob[];
  nextPageToken?: string;
}

export interface EmbeddingResult {
  customId?: string;
  index: number;
  embedding?: number[];
  error?: BatchJobError;
}

export interface EmbeddingsBatchJob {
  name: string;
  state: BatchJobState;
  model: string;
  createTime: string;
  stats?: BatchJobStats;
  results?: EmbeddingResult[];
  outputFileUri?: string;
}

// ========================================
// Batch Job Management
// ========================================

export interface CreateBatchJobOptions {
  requests?: any[];
  inputFilePath?: string;
  model: string;
  displayName?: string;
  outputType?: 'inline' | 'file';
}

/**
 * Create a new batch job
 *
 * @param options - Batch job configuration
 * @returns BatchJob with job ID and initial state
 */
export async function createBatchJob(options: CreateBatchJobOptions): Promise<BatchJob> {
  return invoke('google_batch_create', {
    requests: options.requests,
    inputFilePath: options.inputFilePath,
    model: options.model,
    displayName: options.displayName,
    outputType: options.outputType,
  });
}

/**
 * Get batch job status
 *
 * @param jobName - Batch job name
 * @returns Updated BatchJob with current state
 */
export async function getBatchJob(jobName: string): Promise<BatchJob> {
  return invoke('google_batch_get', { jobName });
}

export interface ListBatchJobsOptions {
  pageSize?: number;
  pageToken?: string;
  filter?: string;
}

/**
 * List all batch jobs
 *
 * @param options - Listing options
 * @returns List of batch jobs and pagination token
 */
export async function listBatchJobs(
  options: ListBatchJobsOptions = {},
): Promise<ListBatchJobsResponse> {
  return invoke('google_batch_list', {
    pageSize: options.pageSize,
    pageToken: options.pageToken,
    filter: options.filter,
  });
}

/**
 * Cancel a running batch job
 *
 * @param jobName - Batch job name
 * @returns Updated BatchJob with CANCELLED state
 */
export async function cancelBatchJob(jobName: string): Promise<BatchJob> {
  return invoke('google_batch_cancel', { jobName });
}

/**
 * Delete a batch job
 *
 * @param jobName - Batch job name
 */
export async function deleteBatchJob(jobName: string): Promise<void> {
  return invoke('google_batch_delete', { jobName });
}

export interface GetBatchResultsOptions {
  jobName: string;
  outputPath?: string;
}

/**
 * Get batch results
 *
 * @param options - Result retrieval options
 * @returns BatchJob with results populated
 */
export async function getBatchResults(options: GetBatchResultsOptions): Promise<BatchJob> {
  return invoke('google_batch_get_results', {
    jobName: options.jobName,
    outputPath: options.outputPath,
  });
}

export interface WaitForCompletionOptions {
  jobName: string;
  pollIntervalSecs?: number;
  maxWaitSecs?: number;
  onProgress?: (job: BatchJob) => void;
}

/**
 * Wait for batch job completion with progress updates
 *
 * @param options - Wait configuration
 * @returns Completed BatchJob
 */
export async function waitForCompletion(options: WaitForCompletionOptions): Promise<BatchJob> {
  const { jobName, pollIntervalSecs = 30, maxWaitSecs = 86400, onProgress } = options;

  const startTime = Date.now();
  const maxDuration = maxWaitSecs * 1000;

  while (true) {
    const job = await getBatchJob(jobName);

    if (onProgress) {
      onProgress(job);
    }

    if (isJobComplete(job.state)) {
      return job;
    }

    if (Date.now() - startTime >= maxDuration) {
      throw new Error(`Batch job timed out after ${maxWaitSecs} seconds`);
    }

    await new Promise((resolve) => setTimeout(resolve, pollIntervalSecs * 1000));
  }
}

/**
 * Check if job state is terminal
 *
 * @param state - Job state
 * @returns True if job is complete
 */
export function isJobComplete(state: BatchJobState): boolean {
  return [
    BatchJobState.SUCCEEDED,
    BatchJobState.FAILED,
    BatchJobState.CANCELLED,
    BatchJobState.EXPIRED,
  ].includes(state);
}

// ========================================
// Embeddings Batch
// ========================================

export interface CreateEmbeddingsBatchOptions {
  texts?: string[];
  inputFilePath?: string;
  model?: string;
  taskType?: string;
  displayName?: string;
}

/**
 * Create embeddings batch job
 *
 * @param options - Embeddings configuration
 * @returns EmbeddingsBatchJob with job ID
 */
export async function createEmbeddingsBatch(
  options: CreateEmbeddingsBatchOptions,
): Promise<EmbeddingsBatchJob> {
  return invoke('google_batch_create_embeddings', {
    texts: options.texts,
    inputFilePath: options.inputFilePath,
    model: options.model || 'gemini-embedding-001',
    taskType: options.taskType,
    displayName: options.displayName,
  });
}

/**
 * Get embeddings batch status
 *
 * @param jobName - Embeddings job name
 * @returns EmbeddingsBatchJob with current state
 */
export async function getEmbeddingsBatch(jobName: string): Promise<EmbeddingsBatchJob> {
  return invoke('google_batch_get_embeddings', { jobName });
}

// ========================================
// Image Generation Batch
// ========================================

export interface CreateImageBatchOptions {
  prompts: string[];
  model: string;
  displayName?: string;
}

/**
 * Create image generation batch job
 *
 * @param options - Image generation configuration
 * @returns BatchJob with job ID
 */
export async function createImageBatch(options: CreateImageBatchOptions): Promise<BatchJob> {
  return invoke('google_batch_create_images', {
    prompts: options.prompts,
    model: options.model,
    displayName: options.displayName,
  });
}

// ========================================
// Utilities
// ========================================

export interface CostEstimate {
  inputCost: number;
  outputCost: number;
  cacheCost: number;
  totalCost: number;
}

/**
 * Calculate batch cost estimate
 *
 * @param model - Model name
 * @param inputTokens - Input token count
 * @param outputTokens - Output token count
 * @param cachedTokens - Cached token count
 * @returns Estimated cost in USD
 */
export async function calculateBatchCost(
  model: string,
  inputTokens: number,
  outputTokens: number,
  cachedTokens: number = 0,
): Promise<number> {
  return invoke('google_batch_calculate_cost', {
    model,
    inputTokens,
    outputTokens,
    cachedTokens,
  });
}

/**
 * Create JSONL file from requests
 *
 * @param requests - Request objects
 * @param outputPath - File path
 */
export async function createJsonlFile(requests: any[], outputPath: string): Promise<void> {
  return invoke('google_batch_create_jsonl', {
    requests,
    outputPath,
  });
}

// ========================================
// Helper Functions
// ========================================

/**
 * Monitor batch job with real-time updates
 *
 * @param jobName - Job name
 * @param callback - Progress callback
 * @param intervalSecs - Poll interval in seconds
 * @returns Promise that resolves when job completes
 */
export async function monitorBatchJob(
  jobName: string,
  callback: (job: BatchJob) => void,
  intervalSecs: number = 30,
): Promise<BatchJob> {
  return waitForCompletion({
    jobName,
    pollIntervalSecs: intervalSecs,
    onProgress: callback,
  });
}

/**
 * Get batch job progress percentage
 *
 * @param job - Batch job
 * @returns Progress percentage (0-100)
 */
export function getBatchProgress(job: BatchJob): number {
  if (!job.stats || job.stats.totalRequests === 0) {
    return 0;
  }

  const completed = job.stats.completedRequests + job.stats.failedRequests;
  return Math.round((completed / job.stats.totalRequests) * 100);
}

/**
 * Get batch job success rate
 *
 * @param job - Batch job
 * @returns Success rate (0-100)
 */
export function getBatchSuccessRate(job: BatchJob): number {
  if (!job.stats || job.stats.totalRequests === 0) {
    return 0;
  }

  const completed = job.stats.completedRequests + job.stats.failedRequests;
  if (completed === 0) {
    return 0;
  }

  return Math.round((job.stats.completedRequests / completed) * 100);
}

/**
 * Extract successful results from batch job
 *
 * @param job - Batch job
 * @returns Array of successful results
 */
export function getSuccessfulResults(job: BatchJob): any[] {
  if (!job.results) {
    return [];
  }

  return job.results.filter((result) => !result.error).map((result) => result.response);
}

/**
 * Extract failed results from batch job
 *
 * @param job - Batch job
 * @returns Array of failed results with errors
 */
export function getFailedResults(job: BatchJob): BatchResult[] {
  if (!job.results) {
    return [];
  }

  return job.results.filter((result) => result.error);
}

/**
 * Estimate time remaining for batch job
 *
 * @param job - Batch job
 * @returns Estimated seconds remaining, or null if unknown
 */
export function estimateTimeRemaining(job: BatchJob): number | null {
  if (!job.stats || !job.createTime) {
    return null;
  }

  const { completedRequests, totalRequests } = job.stats;

  if (completedRequests === 0) {
    return null;
  }

  const elapsedMs = Date.now() - new Date(job.createTime).getTime();
  const avgTimePerRequest = elapsedMs / completedRequests;
  const remainingRequests = totalRequests - completedRequests;

  return Math.round((remainingRequests * avgTimePerRequest) / 1000);
}

/**
 * Format time remaining as human-readable string
 *
 * @param seconds - Seconds
 * @returns Formatted string (e.g., "2h 30m")
 */
export function formatTimeRemaining(seconds: number | null): string {
  if (seconds === null || seconds < 0) {
    return 'Unknown';
  }

  if (seconds < 60) {
    return `${seconds}s`;
  }

  const minutes = Math.floor(seconds / 60);
  if (minutes < 60) {
    return `${minutes}m`;
  }

  const hours = Math.floor(minutes / 60);
  const remainingMinutes = minutes % 60;

  if (hours < 24) {
    return `${hours}h ${remainingMinutes}m`;
  }

  const days = Math.floor(hours / 24);
  const remainingHours = hours % 24;
  return `${days}d ${remainingHours}h`;
}

/**
 * Create a batch job with automatic retry on failure
 *
 * @param options - Batch job options
 * @param maxRetries - Maximum retry attempts
 * @returns Completed batch job
 */
export async function createBatchJobWithRetry(
  options: CreateBatchJobOptions,
  maxRetries: number = 3,
): Promise<BatchJob> {
  let lastError: Error | null = null;

  for (let attempt = 0; attempt < maxRetries; attempt++) {
    try {
      const job = await createBatchJob(options);
      const completed = await waitForCompletion({ jobName: job.name });

      if (completed.state === BatchJobState.SUCCEEDED) {
        return completed;
      }

      // If failed, retry
      lastError = new Error(completed.error?.message || 'Batch job failed');
    } catch (error) {
      lastError = error as Error;
    }

    if (attempt < maxRetries - 1) {
      // Wait before retry (exponential backoff)
      await new Promise((resolve) => setTimeout(resolve, Math.pow(2, attempt) * 1000));
    }
  }

  throw new Error(`Batch job failed after ${maxRetries} attempts: ${lastError?.message}`);
}

// ========================================
// Batch Processing Strategies
// ========================================

/**
 * Process large array in batches
 *
 * @param items - Items to process
 * @param processFn - Function to create request from item
 * @param batchSize - Items per batch
 * @param model - Model to use
 * @returns Array of all results
 */
export async function processBatched<T, R>(
  items: T[],
  processFn: (item: T) => any,
  batchSize: number,
  model: string,
): Promise<R[]> {
  const results: R[] = [];
  const batches = Math.ceil(items.length / batchSize);

  for (let i = 0; i < batches; i++) {
    const start = i * batchSize;
    const end = Math.min(start + batchSize, items.length);
    const batchItems = items.slice(start, end);

    const requests = batchItems.map(processFn);

    const job = await createBatchJob({
      requests,
      model,
      displayName: `Batch ${i + 1}/${batches}`,
    });

    const completed = await waitForCompletion({ jobName: job.name });
    const batchResults = getSuccessfulResults(completed);

    results.push(...batchResults);
  }

  return results;
}

File: apps/desktop/src-tauri/src/sys/commands/google_batch.rs
//! Google Batch API Commands
//!
//! Provides Tauri IPC command shapes for asynchronous large-volume LLM
//! processing through a Google Batch backend when one is connected.
//!
//! This desktop build does not connect directly to Google Batch. Local in-memory
//! preview mode is opt-in via `AGI_ENABLE_LOCAL_IN_MEMORY_GOOGLE_BATCH=1` so IPC
//! calls do not present preview data as a real remote backend.
//!
//! The IPC names and shapes are unchanged so existing UI bindings keep working.

use chrono::Utc;
use once_cell::sync::Lazy;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Mutex;

const LOCAL_PREVIEW_ENV: &str = "AGI_ENABLE_LOCAL_IN_MEMORY_GOOGLE_BATCH";

/// In-memory store for explicitly enabled local preview jobs.
static BATCH_JOBS: Lazy<Mutex<HashMap<String, BatchJob>>> =
    Lazy::new(|| Mutex::new(HashMap::new()));

static EMBEDDINGS_JOBS: Lazy<Mutex<HashMap<String, EmbeddingsBatchJob>>> =
    Lazy::new(|| Mutex::new(HashMap::new()));

fn google_batch_local_preview_enabled() -> bool {
    std::env::var(LOCAL_PREVIEW_ENV)
        .map(|value| matches!(value.as_str(), "1" | "true" | "TRUE" | "yes" | "YES"))
        .unwrap_or(false)
}

fn ensure_google_batch_backend_available() -> Result<(), String> {
    if google_batch_local_preview_enabled() {
        return Ok(());
    }

    Err(format!(
        "Google Batch is not connected in this desktop build. Use the managed cloud batch backend, or set {LOCAL_PREVIEW_ENV}=1 for local in-memory preview mode."
    ))
}

/// Generate a unique job name
fn generate_job_name(prefix: &str) -> String {
    let timestamp = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_millis())
        .unwrap_or(0);
    format!("{}_{}", prefix, timestamp)
}

/// Get current timestamp in ISO 8601 format
fn get_timestamp() -> String {
    Utc::now().to_rfc3339()
}

/// Google AI pricing (input, output, cache) per 1M tokens.
///
/// Sourced from `models.json` via `models_config::get_pricing` so we don't
/// drift from the canonical catalog. The current desktop catalog stores
/// input/output prices only, so cache price is zero until the catalog has an
/// explicit Google cache field.
fn get_model_pricing(model: &str) -> (f64, f64, f64) {
    use crate::core::llm::models_config::get_pricing;
    use crate::core::llm::Provider;

    let pricing = get_pricing(&Provider::Google, model);
    match pricing {
        Some(p) => (p.input_per_million, p.output_per_million, 0.0),
        None => {
            tracing::warn!(
                model = %model,
                "google_batch cost: no pricing in models.json; falling back to (0,0,0)"
            );
            (0.0, 0.0, 0.0)
        }
    }
}

/// Batch job state enum
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "UPPERCASE")]
pub enum BatchJobState {
    Pending,
    Running,
    Succeeded,
    Failed,
    Cancelled,
    Expired,
}

/// Batch job statistics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BatchJobStats {
    pub total_requests: u32,
    pub completed_requests: u32,
    pub failed_requests: u32,
    pub pending_requests: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub total_tokens: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub total_cost: Option<f64>,
}

/// Batch job error
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BatchJobError {
    pub code: i32,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub details: Option<Vec<serde_json::Value>>,
}

/// Batch result entry
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BatchResult {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub custom_id: Option<String>,
    pub index: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub response: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<BatchJobError>,
}

/// Batch job representation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BatchJob {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub display_name: Option<String>,
    pub state: BatchJobState,
    pub model: String,
    pub create_time: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub update_time: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub end_time: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stats: Option<BatchJobStats>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<BatchJobError>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub results: Option<Vec<BatchResult>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output_file_uri: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<serde_json::Value>,
}

/// List batch jobs response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListBatchJobsResponse {
    pub batch_jobs: Vec<BatchJob>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub next_page_token: Option<String>,
}

/// Embedding result entry
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EmbeddingResult {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub custom_id: Option<String>,
    pub index: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub embedding: Option<Vec<f64>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<BatchJobError>,
}

/// Embeddings batch job
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EmbeddingsBatchJob {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub display_name: Option<String>,
    pub state: BatchJobState,
    pub model: String,
    pub create_time: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stats: Option<BatchJobStats>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub results: Option<Vec<EmbeddingResult>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output_file_uri: Option<String>,
}

/// Create a new batch job
/// Returns a BatchJob with PENDING state
#[tauri::command]
pub async fn google_batch_create(
    requests: Option<Vec<serde_json::Value>>,
    _input_file_path: Option<String>,
    model: String,
    display_name: Option<String>,
    _output_type: Option<String>,
) -> Result<BatchJob, String> {
    ensure_google_batch_backend_available()?;

    let request_count = requests.map(|r| r.len() as u32).unwrap_or(0);

    let job = BatchJob {
        name: generate_job_name("batch"),
        display_name,
        state: BatchJobState::Pending,
        model,
        create_time: get_timestamp(),
        update_time: None,
        end_time: None,
        stats: Some(BatchJobStats {
            total_requests: request_count,
            completed_requests: 0,
            failed_requests: 0,
            pending_requests: request_count,
            total_tokens: None,
            total_cost: None,
        }),
        error: None,
        results: None,
        output_file_uri: None,
        metadata: None,
    };

    let job_name = job.name.clone();
    if let Ok(mut jobs) = BATCH_JOBS.lock() {
        jobs.insert(job_name, job.clone());
    }

    Ok(job)
}

/// Get batch job status
/// Returns the BatchJob if found, or an error
#[tauri::command]
pub async fn google_batch_get(job_name: String) -> Result<BatchJob, String> {
    ensure_google_batch_backend_available()?;

    if let Ok(jobs) = BATCH_JOBS.lock() {
        if let Some(job) = jobs.get(&job_name) {
            return Ok(job.clone());
        }
    }
    Err(format!("Batch job '{}' not found", job_name))
}

/// List all batch jobs
/// Returns all stored batch jobs
#[tauri::command]
pub async fn google_batch_list(
    _page_size: Option<u32>,
    _page_token: Option<String>,
    _filter: Option<String>,
) -> Result<ListBatchJobsResponse, String> {
    ensure_google_batch_backend_available()?;

    let jobs = if let Ok(jobs) = BATCH_JOBS.lock() {
        jobs.values().cloned().collect()
    } else {
        Vec::new()
    };

    Ok(ListBatchJobsResponse {
        batch_jobs: jobs,
        next_page_token: None,
    })
}

/// Cancel a running batch job
/// Sets the job state to CANCELLED
#[tauri::command]
pub async fn google_batch_cancel(job_name: String) -> Result<BatchJob, String> {
    ensure_google_batch_backend_available()?;

    if let Ok(mut jobs) = BATCH_JOBS.lock() {
        if let Some(job) = jobs.get_mut(&job_name) {
            job.state = BatchJobState::Cancelled;
            job.update_time = Some(get_timestamp());
            return Ok(job.clone());
        }
    }
    Err(format!("Batch job '{}' not found", job_name))
}

/// Delete a batch job
/// Removes the job from storage
#[tauri::command]
pub async fn google_batch_delete(job_name: String) -> Result<(), String> {
    ensure_google_batch_backend_available()?;

    if let Ok(mut jobs) = BATCH_JOBS.lock() {
        if jobs.remove(&job_name).is_some() {
            return Ok(());
        }
    }
    Err(format!("Batch job '{}' not found", job_name))
}

/// Get batch results
/// Returns the BatchJob with results populated
#[tauri::command]
pub async fn google_batch_get_results(
    job_name: String,
    _output_path: Option<String>,
) -> Result<BatchJob, String> {
    ensure_google_batch_backend_available()?;

    if let Ok(jobs) = BATCH_JOBS.lock() {
        if let Some(job) = jobs.get(&job_name) {
            if matches!(
                job.state,
                BatchJobState::Succeeded | BatchJobState::Failed | BatchJobState::Cancelled
            ) {
                return Ok(job.clone());
            }
            return Ok(job.clone());
        }
    }
    Err(format!("Batch job '{}' not found", job_name))
}

/// Create embeddings batch job
/// Returns an EmbeddingsBatchJob with PENDING state
#[tauri::command]
pub async fn google_batch_create_embeddings(
    texts: Option<Vec<String>>,
    _input_file_path: Option<String>,
    model: Option<String>,
    _task_type: Option<String>,
    display_name: Option<String>,
) -> Result<EmbeddingsBatchJob, String> {
    ensure_google_batch_backend_available()?;

    let text_count = texts.map(|t| t.len() as u32).unwrap_or(0);
    let model = model.ok_or_else(|| {
        "Embedding batch model is required; select it from the configured model catalog."
            .to_string()
    })?;

    let job = EmbeddingsBatchJob {
        name: generate_job_name("embeddings"),
        display_name,
        state: BatchJobState::Pending,
        model,
        create_time: get_timestamp(),
        stats: Some(BatchJobStats {
            total_requests: text_count,
            completed_requests: 0,
            failed_requests: 0,
            pending_requests: text_count,
            total_tokens: None,
            total_cost: None,
        }),
        results: None,
        output_file_uri: None,
    };

    let job_name = job.name.clone();
    if let Ok(mut jobs) = EMBEDDINGS_JOBS.lock() {
        jobs.insert(job_name, job.clone());
    }

    Ok(job)
}

/// Get embeddings batch status
/// Returns the EmbeddingsBatchJob if found
#[tauri::command]
pub async fn google_batch_get_embeddings(job_name: String) -> Result<EmbeddingsBatchJob, String> {
    ensure_google_batch_backend_available()?;

    if let Ok(jobs) = EMBEDDINGS_JOBS.lock() {
        if let Some(job) = jobs.get(&job_name) {
            return Ok(job.clone());
        }
    }
    Err(format!("Embeddings batch job '{}' not found", job_name))
}

/// Create image generation batch job
/// Returns a BatchJob with PENDING state
#[tauri::command]
pub async fn google_batch_create_images(
    prompts: Vec<String>,
    model: String,
    display_name: Option<String>,
) -> Result<BatchJob, String> {
    ensure_google_batch_backend_available()?;

    let prompt_count = prompts.len() as u32;

    let job = BatchJob {
        name: generate_job_name("images"),
        display_name,
        state: BatchJobState::Pending,
        model,
        create_time: get_timestamp(),
        update_time: None,
        end_time: None,
        stats: Some(BatchJobStats {
            total_requests: prompt_count,
            completed_requests: 0,
            failed_requests: 0,
            pending_requests: prompt_count,
            total_tokens: None,
            total_cost: None,
        }),
        error: None,
        results: None,
        output_file_uri: None,
        metadata: None,
    };

    let job_name = job.name.clone();
    if let Ok(mut jobs) = BATCH_JOBS.lock() {
        jobs.insert(job_name, job.clone());
    }

    Ok(job)
}

/// Calculate batch cost estimate
/// Returns estimated cost in USD based on token counts and model pricing
#[tauri::command]
pub async fn google_batch_calculate_cost(
    model: String,
    input_tokens: u64,
    output_tokens: u64,
    cached_tokens: Option<u64>,
) -> Result<f64, String> {
    let (input_price, output_price, cache_price) = get_model_pricing(&model);
    if cached_tokens.unwrap_or(0) > 0 && cache_price == 0.0 {
        return Err(
            "Google cache-token pricing is not present in the desktop model catalog; cannot estimate cached token cost."
                .to_string(),
        );
    }

    // Calculate costs (prices are per 1M tokens)
    let input_cost = (input_tokens as f64 / 1_000_000.0) * input_price;
    let output_cost = (output_tokens as f64 / 1_000_000.0) * output_price;
    let cache_cost = cached_tokens
        .map(|ct| (ct as f64 / 1_000_000.0) * cache_price)
        .unwrap_or(0.0);

    let total_cost = input_cost + output_cost + cache_cost;

    Ok(total_cost)
}

/// Create JSONL file from requests
/// Writes requests to a JSONL file at the specified path
#[tauri::command]
pub async fn google_batch_create_jsonl(
    requests: Vec<serde_json::Value>,
    output_path: String,
) -> Result<(), String> {
    let safe_output_path = crate::sys::commands::file_ops::validate_path_security(&output_path)?;

    use std::fs::File;
    use std::io::Write;

    let file =
        File::create(&safe_output_path).map_err(|e| format!("Failed to create file: {}", e))?;

    let mut writer = std::io::BufWriter::new(file);

    for request in requests {
        let json_str = serde_json::to_string(&request)
            .map_err(|e| format!("Failed to serialize request: {}", e))?;
        writeln!(writer, "{}", json_str).map_err(|e| format!("Failed to write to file: {}", e))?;
    }

    writer
        .flush()
        .map_err(|e| format!("Failed to flush writer: {}", e))?;

    Ok(())
}

/// Compatibility command the frontend can poll to decide whether to render a
/// Google Batch unavailable/local-preview banner. The command name is retained
/// for existing UI bindings.
#[tauri::command]
pub async fn google_batch_is_beta_stub() -> Result<bool, String> {
    Ok(!google_batch_local_preview_enabled())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pricing_is_sourced_from_catalog_for_current_gemini_models() {
        // gemini-3.5-flash-lite is in models.json at $0.30/$2.50 per 1M.
        let (input, output, cache) = get_model_pricing("gemini-3.5-flash-lite");
        assert_eq!(input, 0.3);
        assert_eq!(output, 2.5);
        assert_eq!(cache, 0.0);
    }

    #[test]
    fn pricing_returns_zeros_for_unknown_model_name() {
        let (input, output, cache) = get_model_pricing("not-a-real-model-id");
        // Falls back to provider default if catalog miss.
        assert!(input >= 0.0);
        assert!(output >= 0.0);
        assert_eq!(cache, 0.0);
    }
}
